File: NumberedCard.java
public class NumberedCard extends Card {

    private long id;
    private static long counter = 0;

    public NumberedCard(Card card) {
        super(card.getSuit(), card.getRank());
        id = counter++;
    }

    public NumberedCard(Card card, long id) {
        super(card.getSuit(), card.getRank());
        this.id = id;
    }

    @Override
    public String toString() {
        return super.toString() + "$" + id;
    }

    @Override
    public boolean equals(Object o) {
        return super.equals(o) && ((NumberedCard) o).id == this.id;
    }
}

File: src/main/java/com/isaac/games/NumberedCard.kt
package com.isaac.games

class NumberedCard : Card {

    private var id: Long = 0

    constructor(card: Card) : super(card.suit, card.rank) {
        id = counter++
    }

    constructor(card: Card, id: Long) : super(card.suit, card.rank) {
        this.id = id
    }

    override fun toString(): String {
        return super.toString() + "$" + id
    }

    override fun equals(o: Any?): Boolean {
        return super.equals(o) && (o as NumberedCard).id == this.id
    }

    companion object {
        private var counter: Long = 0
    }
}